src/providers/postgres/qgspostgresprovidermetadatautils.cpp

```diff
@@ -0,0 +1,365 @@
+  }
+  else
+  {
+    // Unsupported!
+    return false;
+  }
+
+  QgsPostgresConn *conn = QgsPostgresConn::connectDb( dsUri.connectionInfo( false ), false );
+  if ( !conn )
+  {
+    errorMessage = QObject::tr( "Connection to database failed" );
+    return false;
+  }
+
+  if ( dsUri.database().isEmpty() ) // typically when a service file is used
+  {
+    dsUri.setDatabase( conn->currentDatabase() );
+  }
+
+  // Try to load metadata
+  QString schemaName { dsUri.schema().isEmpty() ? QStringLiteral( "public" ) : dsUri.schema() };
+  const QString schemaQuery = QStringLiteral( "SELECT table_schema FROM information_schema.tables WHERE table_name = 'qgis_layer_metadata'" );
+  QgsPostgresResult res( conn->LoggedPQexec( "QgsPostgresProviderMetadataUtils", schemaQuery ) );
+  const bool metadataTableFound { res.PQntuples( ) > 0 };
+  if ( metadataTableFound )
+  {
+    schemaName = res.PQgetvalue( 0, 0 ) ;
+  }
+  else
+  {
+    QgsPostgresResult res( conn->LoggedPQexec( QStringLiteral( "QgsPostgresProviderMetadataUtils" ),
+                           QStringLiteral( R"SQL(
+            CREATE TABLE %1.qgis_layer_metadata (
+              id SERIAL PRIMARY KEY
+              ,f_table_catalog VARCHAR NOT NULL
+              ,f_table_schema VARCHAR NOT NULL
+              ,f_table_name VARCHAR NOT NULL
+              ,f_geometry_column VARCHAR
+              ,identifier TEXT NOT NULL
+              ,title TEXT NOT NULL
+              ,abstract TEXT
+              ,geometry_type VARCHAR
+              ,extent GEOMETRY(POLYGON, 4326)
+              ,crs VARCHAR
+              ,layer_type VARCHAR NOT NULL
+              ,qmd XML NOT NULL
+              ,owner VARCHAR(63) DEFAULT CURRENT_USER
+              ,update_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+              UNIQUE (f_table_catalog, f_table_schema, f_table_name, f_geometry_column, geometry_type, crs, layer_type)
+            )
+          )SQL" ).arg( QgsPostgresConn::quotedIdentifier( schemaName ) ) ) );
+    if ( res.PQresultStatus() != PGRES_COMMAND_OK )
+    {
+      errorMessage = QObject::tr( "Unable to save layer metadata. It's not possible to create the destination table on the database. Maybe this is due to table permissions (user=%1). Please contact your database admin" ).arg( dsUri.username() );
+      conn->unref();
+      return false;
+    }
+  }
+
+  const QString wkbTypeString = QgsWkbTypes::geometryDisplayString( QgsWkbTypes::geometryType( dsUri.wkbType() ) );
+
+  const QgsCoordinateReferenceSystem metadataCrs { metadata.crs() };
+  QgsCoordinateReferenceSystem destCrs {QgsCoordinateReferenceSystem::fromEpsgId( 4326 ) };
+  QgsRectangle extents;
+
+  const auto cExtents { metadata.extent().spatialExtents() };
+  for ( const auto &ext : std::as_const( cExtents ) )
+  {
+    QgsRectangle bbox {  ext.bounds.toRectangle()  };
+    QgsCoordinateTransform ct { ext.extentCrs, QgsCoordinateReferenceSystem::fromEpsgId( 4326 ), QgsProject::instance()->transformContext() };
```

Comment:
No QgsProject::instance() here

Reply:
I cannot pass the context here, it would be an API break, QgsProviderMetadata::saveLayerMetadata was added in 3.20
I can create a default transform context if the project instance is a no-go

Reply:
A default transform context is acceptable to me -- we don't need meter level accuracy here. (Maybe just add a comment of that nature when you create the context, so that there's low risk someone will copy/paste that code in future into a part of QGIS where accuracy IS important!).